diff --git a/src/libsodium/sodium/runtime.c b/src/libsodium/sodium/runtime.c
index 2251da9a..0da45405 100644
--- a/src/libsodium/sodium/runtime.c
+++ b/src/libsodium/sodium/runtime.c
@@ -182,7 +182,12 @@ _cpuid(unsigned int cpu_info[4U], const unsigned int cpu_info_type)
                            "=c"(cpu_info[2]), "=d"(cpu_info[3])
                          : "0"(cpu_info_type), "2"(0U));
 # endif
-#else
+#elif defined(__llir__)
+    __asm__ __volatile__
+        ( "x86_cpuid.i32.i32.i32.i32 %0, %1, %2, %3, %4, %5"
+        : "=r"(cpu_info[0]), "=r"(cpu_info[1]), "=r"(cpu_info[2]), "=r"(cpu_info[3])
+        : "r"(cpu_info_type), "r"(0U)
+        );
     (void) cpu_info_type;
     cpu_info[0] = cpu_info[1] = cpu_info[2] = cpu_info[3] = 0;
 #endif
