diff --git a/ev.c b/ev.c
index ec212a1..1a7a442 100644
--- a/ev.c
+++ b/ev.c
@@ -795,9 +795,15 @@ struct signalfd_siginfo
       #define ECB_MEMORY_FENCE_ACQUIRE __asm__ __volatile__ (""                        : : : "memory")
       #define ECB_MEMORY_FENCE_RELEASE __asm__ __volatile__ (""                        : : : "memory")
     #elif ECB_GCC_AMD64
-      #define ECB_MEMORY_FENCE         __asm__ __volatile__ ("mfence"   : : : "memory")
-      #define ECB_MEMORY_FENCE_ACQUIRE __asm__ __volatile__ (""         : : : "memory")
-      #define ECB_MEMORY_FENCE_RELEASE __asm__ __volatile__ (""         : : : "memory")
+      #if defined(__llir__)
+        #define ECB_MEMORY_FENCE         __asm__ __volatile__ ("x86_m_fence"   : : : "memory")
+        #define ECB_MEMORY_FENCE_ACQUIRE __asm__ __volatile__ (""              : : : "memory")
+        #define ECB_MEMORY_FENCE_RELEASE __asm__ __volatile__ (""              : : : "memory")
+      #else
+        #define ECB_MEMORY_FENCE         __asm__ __volatile__ ("mfence"   : : : "memory")
+        #define ECB_MEMORY_FENCE_ACQUIRE __asm__ __volatile__ (""         : : : "memory")
+        #define ECB_MEMORY_FENCE_RELEASE __asm__ __volatile__ (""         : : : "memory")
+      #endif
     #elif __powerpc__ || __ppc__ || __powerpc64__ || __ppc64__
       #define ECB_MEMORY_FENCE         __asm__ __volatile__ ("sync"     : : : "memory")
     #elif defined __ARM_ARCH_2__ \
