diff --git a/config/cfg.ml b/config/cfg.ml
index c880380..8ec9758 100644
--- a/config/cfg.ml
+++ b/config/cfg.ml
@@ -16,12 +16,12 @@ let _ =
   in
   let accelerate_flags =
     match arch with
-    | `x86_64 -> [ "-DACCELERATE"; "-mssse3"; "-maes"; "-mpclmul" ]
+    | `x86_64 -> [ ]
     | _ -> []
   in
   let ent_flags =
     match arch with
-    | `x86_64 | `x86 -> [ "-DENTROPY"; "-mrdrnd"; "-mrdseed" ]
+    | `x86_64 | `x86 -> [ ]
     | _ -> []
   in
   let flags = std_flags @ ent_flags @ accelerate_flags in
