diff --git a/config/cfg.ml b/config/cfg.ml
index c880380..9ea5ea7 100644
--- a/config/cfg.ml
+++ b/config/cfg.ml
@@ -7,20 +7,23 @@ let _ =
       Configurator.V1.C_define.import
         c
         ~includes:[]
-        [("__x86_64__", Switch); ("__i386__", Switch)]
+        [("__llir__", Switch); ("__x86_64__", Switch); ("__i386__", Switch)]
     in
     match defines with
-    | (_, Switch true) :: _ -> `x86_64
-    | _ :: (_, Switch true) :: _ -> `x86
+    | (_, Switch true) :: (_, Switch true) :: _ -> `llir_x86_64
+    | _ :: (_, Switch true) :: _ -> `x86_64
+    | _ :: _ :: (_, Switch true) :: _ -> `x86
     | _ -> `unknown
   in
   let accelerate_flags =
     match arch with
+    | `llir_x86_64 -> []
     | `x86_64 -> [ "-DACCELERATE"; "-mssse3"; "-maes"; "-mpclmul" ]
     | _ -> []
   in
   let ent_flags =
     match arch with
+    | `llir_x86_64 -> []
     | `x86_64 | `x86 -> [ "-DENTROPY"; "-mrdrnd"; "-mrdseed" ]
     | _ -> []
   in
