diff --git a/include/verilatedos.h b/include/verilatedos.h
index 9a1f50a7..fff8459c 100644
--- a/include/verilatedos.h
+++ b/include/verilatedos.h
@@ -332,6 +332,7 @@ typedef signed __int32 ssize_t;  ///< signed size_t; returned from read()
 # include <inttypes.h>  // Solaris
 # include <stdint.h>  // Linux and most flavors
 # include <sys/types.h>  // __WORDSIZE
+# include <sys/reg.h>
 # include <unistd.h>  // ssize_t
 typedef char vlsint8_t;  ///< 8-bit signed type
 typedef uint8_t vluint8_t;  ///< 8-bit unsigned type
