diff --git a/lib/barrier.h b/lib/barrier.h
index 4a9e6b1..a34392f 100644
--- a/lib/barrier.h
+++ b/lib/barrier.h
@@ -13,9 +13,15 @@
 #define xen_rmb() xen_barrier()
 #define xen_wmb() xen_barrier()
 #elif defined(__x86_64__)
-#define xen_mb()  asm volatile ( "mfence" : : : "memory")
-#define xen_rmb() xen_barrier()
-#define xen_wmb() xen_barrier()
+#  if defined(__llir__)
+#  define xen_mb()  asm volatile ( "x86_m_fence" : : : "memory")
+#  define xen_rmb() xen_barrier()
+#  define xen_wmb() xen_barrier()
+#else
+#  define xen_mb()  asm volatile ( "mfence" : : : "memory")
+#  define xen_rmb() xen_barrier()
+#  define xen_wmb() xen_barrier()
+#endif
 #elif defined(__arm__)
 # ifndef _M_ARM
 #define xen_mb()   {}
